### PR TITLE
metrics/params: support reading from cached files and directories, use repo-relative path consistently

### PR DESCRIPTION
**Pull request by skshetry**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Closes https://github.com/iterative/dvc/issues/9452.
Closes https://github.com/iterative/dvc/issues/9478.
Also addresses https://github.com/iterative/studio/issues/4912.

TODO
- [x] Add tests for cached files read and directories.

### CODE REVIEW

**Comment by skshetry** [comment]
@dberenbaum, I commented here https://github.com/iterative/dvc/issues/9588#issuecomment-1706202261.

TLDR: parameter dependencies are granular, so I am not sure it's worth the complications. :)

**Comment by skshetry** [comment]
@dberenbaum, it's not needed to close https://github.com/iterative/studio/issues/4912. This should be enough.

File: dvc/repo/params/show.py [hunk 1]
```diff
@@ -117,74 +95,101 @@ def _collect_vars(repo, params, stages=None) -> Dict:
...
+                deps_only=deps_only,
+                on_error=on_error,
+            )
+            res[rev] = Result(data=params)
+        except Exception as exc:  # noqa: BLE001 # pylint:disable=broad-exception-caught
+            if on_error == "raise":
+                raise
+            logger.warning("failed to load params in revision %r, %s", rev, str(exc))
+            if on_error == "return":
+                res[rev] = Result(error=exc)
```
Comment (skshetry) [inline dvc/repo/params/show.py:189]: I am moving away from `onerror` callback. Moving forward, I'd prefer to have a tighter control, and have explicit places (and fewer places) where we ignore/suppress exception.

I am okay with `onerror` callback as a way to decide whether or not to ignore an exception as an exception handler, but I don't like it as a way to change return type.

File: dvc/repo/params/show.py [hunk 2]
```diff
@@ -117,74 +95,101 @@ def _collect_vars(repo, params, stages=None) -> Dict:
...
+        if not isinstance(result, Exception):
+            data.update({repo_os_path: FileResult(data=result)})
+            continue
+
+        if on_error == "raise":
+            raise result
+        if on_error == "return":
+            data.update({repo_os_path: FileResult(error=result)})
+
+    # vars_params = _collect_vars(repo, params, stages=stages)
```
Comment (skshetry) [inline dvc/repo/params/show.py:141]: We used to show `params` from the variables used in the templating (other than from `params.yaml` file). But determining the true source has been challenging (especially when the paths have to be consistent), and merging all of them is challenging. The implementation is incomplete too (it is skipping any file that is already read, but params are granular that it's not enough).

We have top-level params now, that this does not feel necessary or useful.

With templating variable, i meant this:
```yaml
# params1.yaml
foo: foobar

# dvc.yaml
vars:
- params1.yaml

stages:
  stage1:
    cmd: echo ${foobar}
```

```console
$ dvc params diff
Path          Param    HEAD    workspace                              
params1.yaml  foo      -       foobar
```

But the implementation is incomplete, this was only shown when the file is not specified anywhere as params.

File: dvc/repo/params/show.py (hunk 1 above)
Comment (dberenbaum) [reply dvc/repo/params/show.py:162]: Should we expect anything to fail that wouldn't fail before?

File: dvc/repo/params/show.py (hunk 2 above)
Comment (dberenbaum) [reply dvc/repo/params/show.py]: 🤔 This seems like it will be surprising for people who are using features like dict unpacking. Would they now need to explicitly list all of the unpacked args in params?

Comment (skshetry) [reply dvc/repo/params/show.py]: Brought that back.

File: dvc/repo/params/show.py (hunk 1 above)
Comment (skshetry) [reply dvc/repo/params/show.py:200]: Nothing should fail. We are still suppressing those.

**Comment by daavoo** [comment]
Was this tested against Studio?

**Comment by 0x2b3bfa0** [comment]
* https://github.com/iterative/dvc/issues/10177

**Comment by 0x2b3bfa0** [comment]
> — _Was this tested against Studio?_
> — _Nothing changed that affects Studio._

I want a T-shirt with that. 😄

**Comment by skshetry** [comment]
That wasn't introduced here. This PR only lifted the method to a free function.

https://github.com/iterative/dvc/pull/9909/files#diff-166cafd50d9e2d242cc8cf2b9fce6231227513e7342b705fa25c27abb24b0fa4L97

**Comment by 0x2b3bfa0** [comment]
> _That wasn't introduced here._

As per `git bisect` results, this pull request introduced the issue. Even though the code was already there, it wasn't being called when evaluating the #10177 minimal example. 🤔 

> _It was introduced very long ago (https://github.com/iterative/dvc/pull/5550), and it's an honest mistake that anyone could make._

Indeed!

**Comment by skshetry** [comment]
It is most likely because after this PR, we merge all of the keypaths for the same file from different stages to avoid reading multiple times. The keypaths you have in a single stage is likely not enough to trigger that.
